apps/src/lib/ui/accounts/ChangeEmailController.js

```diff
@@ -0,0 +1,119 @@
+import React from 'react';
+import ReactDOM from "react-dom";
+import color from '../../../util/color';
+import ChangeEmailModal from './ChangeEmailModal';
+import {hashEmail} from '../../../code-studio/hashEmail';
+
+export const ENCRYPTED_EMAIL_PLACEHOLDER = '***encrypted***';
+
+/**
+ * This controller submits email changes to dashboard using a hidden
+ * Rails-generated form.  It expects a jQuery wrapper for that form to be
+ * provided on construction.
+ *
+ * When the user submits an email change, this controller loads the relevant
+ * information into the hidden Rails form and calls submit(). Rails injects
+ * all the JavaScript needed for the form to submit via AJAX with all the
+ * appropriate validation tokens, etc.  The controller subscribes to events
+ * emitted by the Rails helper JavaScript to detect success or errors.
+ *
+ * Read more:
+ * http://guides.rubyonrails.org/working_with_javascript_in_rails.html#rails-ujs-event-handlers
+ * https://github.com/rails/jquery-ujs
+ */
+export default class ChangeEmailController {
+  /**
+   * @param {jQuery} form
+   * @param {jQuery} link
+   * @param {jQuery} displayedUserEmail
+   * @param {jQuery} userAge
+   * @param {function(newEmail:string, newHashedEmail:string)} emailChangedCallback
+   */
+  constructor({form, link, displayedUserEmail, userAge, emailChangedCallback}) {
+    this.form = form;
+    this.displayedUserEmail = displayedUserEmail;
+    this.userAge = userAge;
+    this.emailChangedCallback = emailChangedCallback;
+    link.click(this.showChangeEmailModal);
+  }
+
+  showChangeEmailModal = () => {
+    const userHashedEmail = this.form.find('#change-email-modal_user_hashed_email').val();
+    const handleSubmit = (values) => (
+      this.submitEmailChange(values)
+        .then(this.onEmailChanged)
+    );
+
+    this.mountPoint = document.createElement('div');
+    document.body.appendChild(this.mountPoint);
+    ReactDOM.render(
+      <ChangeEmailModal
+        handleSubmit={handleSubmit}
+        handleCancel={this.hideChangeEmailModal}
+        currentHashedEmail={userHashedEmail}
+      />,
+      this.mountPoint
+    );
+  };
+
+  hideChangeEmailModal = () => {
+    if (this.mountPoint) {
```

Comment:
Is for idempotency? Should we have a similar inverse check in `showChangeEmailModal`?

Reply:
That's a good idea, I didn't think of that.  Making _hide_ idempotent was useful in tests to guarantee test cleanup, but it'd make sense to have parallel idempotency in _show_.  I'll take care of this now!